File: src/ios/AppDelegate+FirebasePlugin.m
#import "AppDelegate+FirebasePlugin.h"
#import "FirebasePlugin.h"
#import <objc/runtime.h>

#if defined(__IPHONE_10_0) && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_10_0
  @import UserNotifications;

  // Implement UNUserNotificationCenterDelegate to receive display notification via APNS for devices
  // running iOS 10 and above. Implement FIRMessagingDelegate to receive data message via FCM for
  // devices running iOS 10 and above.
  @interface AppDelegate () <UNUserNotificationCenterDelegate>
  @end
#endif

#define kApplicationInBackgroundKey @"applicationInBackground"
#define kDelegateKey @"delegate"

@implementation AppDelegate (FirebasePlugin)

+ (void)load {
    Method original = class_getInstanceMethod(self, @selector(application:didFinishLaunchingWithOptions:));
    Method swizzled = class_getInstanceMethod(self, @selector(identity_application:didFinishLaunchingWithOptions:));
    method_exchangeImplementations(original, swizzled);
}

- (BOOL)identity_application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: didFinishLaunchingWithOptions called");
    
    // MOCK: Simulate Firebase initialization without actual Firebase calls
    [self mockInitializeFirebase];
    
    return [self identity_application:application didFinishLaunchingWithOptions:launchOptions];
}

- (void)mockInitializeFirebase {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: Initializing Firebase in mock mode");
    
    // MOCK: Set up notification observers without actual Firebase calls
    [self setupMockNotificationObservers];
}

- (void)setupMockNotificationObservers {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: Setting up mock notification observers");
    
    // MOCK: Don't actually register for notifications, just log
    #if defined(__IPHONE_10_0) && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_10_0
        NSLog(@"AppDelegate+FirebasePlugin MOCK: Would set UNUserNotificationCenter delegate");
    #endif
    
    // MOCK: Don't actually register for token refresh, just log
    NSLog(@"AppDelegate+FirebasePlugin MOCK: Would register for token refresh notifications");
}

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: didRegisterForRemoteNotificationsWithDeviceToken called");
    // MOCK: Don't actually set APNS token, just log
}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: didFailToRegisterForRemoteNotificationsWithError called");
    // MOCK: Just log the error, don't handle it
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: didReceiveRemoteNotification called");
    // MOCK: Don't actually handle the notification, just log
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: didReceiveRemoteNotification with completion handler called");
    // MOCK: Always return no new data
    completionHandler(UIBackgroundFetchResultNoData);
}

- (void)tokenRefreshNotification:(NSNotification *)notification {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: tokenRefreshNotification called");
    // MOCK: Don't actually get token, just log
}

- (void)messaging:(id)messaging didReceiveMessage:(id)remoteMessage {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: messaging:didReceiveMessage called");
    // MOCK: Don't actually handle the message, just log
}

- (void)applicationReceivedRemoteMessage:(id)remoteMessage {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: applicationReceivedRemoteMessage called");
    // MOCK: Don't actually handle the message, just log
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: applicationDidBecomeActive called");
    // MOCK: Don't actually connect to messaging, just log
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
    NSLog(@"AppDelegate+FirebasePlugin MOCK: applicationDidEnterBackground called");
    // MOCK: Don't actually disconnect from messaging, just log
}

@end

File: src/ios/AppDelegate+FirebasePlugin.swift
import Foundation
import UIKit
import Firebase
import FirebaseMessaging
import UserNotifications
import ObjectiveC

// MARK: - AppDelegate Extension for Firebase Plugin
extension AppDelegate {
    
    // MARK: - Associated Objects Keys
    private static let kApplicationInBackgroundKey = "applicationInBackground"
    private static let kDelegateKey = "delegate"
    
    // MARK: - Associated Objects Properties
    @objc var applicationInBackground: NSNumber? {
        get { return objc_getAssociatedObject(self, &AppDelegate.kApplicationInBackgroundKey) as? NSNumber }
        set { objc_setAssociatedObject(self, &AppDelegate.kApplicationInBackgroundKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC) }
    }
    
    // MARK: - Method Swizzling
    @objc static func swizzleMethods() {
        let originalSelector = #selector(UIApplicationDelegate.application(_:didFinishLaunchingWithOptions:))
        let swizzledSelector = #selector(AppDelegate.application(_:swizzledDidFinishLaunchingWithOptions:))
        
        guard let originalMethod = class_getInstanceMethod(AppDelegate.self, originalSelector),
              let swizzledMethod = class_getInstanceMethod(AppDelegate.self, swizzledSelector) else {
            return
        }
        
        let didAddMethod = class_addMethod(AppDelegate.self, originalSelector,
                                         method_getImplementation(swizzledMethod),
                                         method_getTypeEncoding(swizzledMethod))
        
        if didAddMethod {
            class_replaceMethod(AppDelegate.self, swizzledSelector,
                              method_getImplementation(originalMethod),
                              method_getTypeEncoding(originalMethod))
        } else {
            method_exchangeImplementations(originalMethod, swizzledMethod)
        }
    }

    @objc func application(_ application: UIApplication, swizzledDidFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        print("FirebasePlugin - AppDelegate swizzled didFinishLaunchingWithOptions")
        
        // Initialize Firebase
        FirebaseApp.configure()
        
        // Set messaging delegate
        Messaging.messaging().delegate = self
        
        // Set UNUserNotificationCenter delegate
        UNUserNotificationCenter.current().delegate = self
        
        // Register for remote notifications
        application.registerForRemoteNotifications()
        
        // Call original method
        let handled = self.application(application, swizzledDidFinishLaunchingWithOptions: launchOptions)
        return handled
    }
}

// MARK: - MessagingDelegate
extension AppDelegate: MessagingDelegate {
    func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
        print("FirebasePlugin - FCM registration token: \(fcmToken ?? "nil")")
        
        // Notify JavaScript if callback is registered
        if let plugin = FirebasePlugin.firebasePlugin,
           let callbackId = plugin.tokenRefreshCallbackId {
            let result = CDVPluginResult(status: .ok, messageAs: fcmToken)
            plugin.commandDelegate.send(result, callbackId: callbackId)
        }
    }
}

// MARK: - UNUserNotificationCenterDelegate
extension AppDelegate: UNUserNotificationCenterDelegate {
    func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        print("FirebasePlugin - Will present notification: \(notification.request.content.userInfo)")
        
        // Show notification even when app is in foreground
        completionHandler([.alert, .badge, .sound])
    }
    
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        print("FirebasePlugin - Did receive notification response: \(response.notification.request.content.userInfo)")
        
        // Notify JavaScript if callback is registered
        if let plugin = FirebasePlugin.firebasePlugin,
           let callbackId = plugin.notificationCallbackId {
            let userInfo = response.notification.request.content.userInfo
            let result = CDVPluginResult(status: .ok, messageAs: userInfo)
            plugin.commandDelegate.send(result, callbackId: callbackId)
        }
        
        completionHandler()
    }
}

// MARK: - Static reference to FirebasePlugin
extension FirebasePlugin {
    static var firebasePlugin: FirebasePlugin?
}
